keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1294 @@
+  request has been issued, denied, or failed.
+
+The noderestriction admission plugin checks the consistency of the information in created PodCertificateRequest objects:
+* The namespace, pod name/UID, service account name/UID, and node name/UID must
+  correspond to a real pod that currently exists in the cluster.
+* The pod should be in the Pending or Running phases.
+* The PodCertificateRequest must have been created by the user
+  `system:node:foo`, where `foo` is the node named in the PodCertificateRequest
+  spec.
+
+This upholds the Kubernetes node isolation guarantee by ensuring that
+compromising a single node only results in compromises of workloads that have
+actually been scheduled onto that node.
+
+During alpha, the ability to store PodCertificate objects (and the node
+restriction admission logic) will be gated by the PodCertificateRequests feature
+gate.
+
+<<[UNRESOLVED]>>
+Should RSA2048 be offered?  @ahmedtd inclined to say no.  Are there any other key types that should be offered?
+<<[/UNRESOLVED]>>
+
+#### Why a New Type?
+
+The following constraints led to the adoption of a new API type, as opposed to extending CertificateSigningRequest:
+1. We wanted the pod details to be legible without having to parse an ASN.1 CSR.
+2. We wanted the pod details (especially the associated node name) to be
+  available for use in an informer selector.  While these could be surfaced as labels on CertificateSigningRequest objects, labels are immutable.
+3. We wanted pod details to be immutable.
+4. We wanted to avoid putting Pod Certificate-specific fields in the generic
+  CertificateSigningRequest.
+5. Avoid confusing existing signer implementations with pod information they
+   don't understand.
+
+The available options were:
+
+* Label CertificateSigningRequests with node/pod labels by convention.  This didn't meet goal 3.
+* Hoist pod fields into CSR spec.  This didn't meet goal 4.
+* Embed pod details into CSR spec.request field.  This didn't meet goals 1 and 2.
+* New dedicated PodCertificateRequest type. Meets all goals.
+
+#### API Definition
+
+```go
+
+// PodCertificateRequest encodes a pod requesting a certificate from a given
+// signer.
+//
+// Kubelets use this API to implement podCertificate projected volumes
+type PodCertificateRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec contains the details about the certificate being requested.
+	Spec PodCertificateRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status contains the issued certificate, and a standard set of conditions.
+	// +optional
+	Status PodCertificateRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// PodCertificateRequestSpec describes the certificate request.  All fields are
+// immutable after creation.
+type PodCertificateRequestSpec struct {
```

Comment:
Consider whether a pod request for a certificate should be able to express a hint for certificate lifetime, similar to https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2784-csr-duration#motivation

A minimum (maybe the same as CSR of 10 minutes) to avoid the kubelet churning on expiring certificates could make sense.